# BÀI TẬP: TITANIC
**Nguồn:** kaggle.com/c/titanic (891 dòng)


In [1]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt, seaborn as sns
from pathlib import Path
from scipy import stats

sns.set_style('whitegrid')

csv_path = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'

df = pd.read_csv(csv_path)
print('Loaded from:', csv_path)
df.head()

Loaded from: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


---
# PHẦN A — DATA PROFILING
## A.1. Data size, column names, data types

In [2]:
# TODO
print("Kích thước dữ liệu (rows, columns):", df.shape)
print("\nDanh sách các cột:")
print(df.columns.tolist())
print("\nKiểu dữ liệu chi tiết:")
print(df.dtypes)
print("\nThông tin tổng quan:")
df.info()

Kích thước dữ liệu (rows, columns): (891, 15)

Danh sách các cột:
['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']

Kiểu dữ liệu chi tiết:
survived         int64
pclass           int64
sex                str
age            float64
sibsp            int64
parch            int64
fare           float64
embarked           str
class              str
who                str
adult_male        bool
deck               str
embark_town        str
alive              str
alone             bool
dtype: object

Thông tin tổng quan:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    str    
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int

## A.2. Missing values & Duplicate data

In [3]:
# TODO
print(df.isnull().sum())

duplicate_count = df.duplicated().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


## A.3. Invalid values

In [4]:
# TODO
print(df[['age', 'sibsp', 'parch', 'fare']].describe())

invalid_age_fare = df[(df['age'] < 0) | (df['fare'] < 0)]

              age       sibsp       parch        fare
count  714.000000  891.000000  891.000000  891.000000
mean    29.699118    0.523008    0.381594   32.204208
std     14.526497    1.102743    0.806057   49.693429
min      0.420000    0.000000    0.000000    0.000000
25%     20.125000    0.000000    0.000000    7.910400
50%     28.000000    0.000000    0.000000   14.454200
75%     38.000000    1.000000    0.000000   31.000000
max     80.000000    8.000000    6.000000  512.329200


## A.4. Create a new column
Tạo cột `family_size` = sibsp + parch + 1.

In [8]:
# TODO
df['family_size'] = df['sibsp'] + df['parch'] + 1
print(df['family_size'].head())


0    2
1    2
2    1
3    2
4    1
Name: family_size, dtype: int64


---
# PHẦN B — DESCRIPTIVE STATISTICS
## Group 1 — Central Tendency

In [10]:
# TODO
cols = ['age', 'fare', 'family_size']

print("Giá trị trung bình (Mean):")
print(df[cols].mean())

print("\nGiá trị trung vị (Median):")
print(df[cols].median())

print("\nGiá trị xuất hiện nhiều nhất :")
print(df[cols].mode())

Giá trị trung bình (Mean):
age            29.699118
fare           32.204208
family_size     1.904602
dtype: float64

Giá trị trung vị (Median):
age            28.0000
fare           14.4542
family_size     1.0000
dtype: float64

Giá trị xuất hiện nhiều nhất :
    age  fare  family_size
0  24.0  8.05            1


## Group 2 — Dispersion

In [11]:
# TODO
print("Độ lệch chuẩn (Standard Deviation):")
print(df[cols].std())

print("\nPhương sai (Variance):")
print(df[cols].var())

print("\nKhoảng biến thiên (Range = Max - Min):")
print(df[cols].max() - df[cols].min())

print("\nKhoảng tứ phân vị (IQR = Q3 - Q1):")
print(df[cols].quantile(0.75) - df[cols].quantile(0.25))

Độ lệch chuẩn (Standard Deviation):
age            14.526497
fare           49.693429
family_size     1.613459
dtype: float64

Phương sai (Variance):
age             211.019125
fare           2469.436846
family_size       2.603248
dtype: float64

Khoảng biến thiên (Range = Max - Min):
age             79.5800
fare           512.3292
family_size     10.0000
dtype: float64

Khoảng tứ phân vị (IQR = Q3 - Q1):
age            17.8750
fare           23.0896
family_size     1.0000
dtype: float64


## Group 3 — Location and Shape

In [12]:
# TODO
print("Độ lệch (Skewness):")
print(df[cols].skew())

print("\nĐộ nhọn (Kurtosis):")
print(df[cols].kurt())

print("\nCác mức phân vị (Q1=25%, Q2=50%, Q3=75%):")
print(df[cols].quantile([0.25, 0.5, 0.75]))

Độ lệch (Skewness):
age            0.389108
fare           4.787317
family_size    2.727441
dtype: float64

Độ nhọn (Kurtosis):
age             0.178274
fare           33.398141
family_size     9.159666
dtype: float64

Các mức phân vị (Q1=25%, Q2=50%, Q3=75%):
         age     fare  family_size
0.25  20.125   7.9104          1.0
0.50  28.000  14.4542          1.0
0.75  38.000  31.0000          2.0


---
# PHẦN C — DEFINE THE QUESTION

## Câu hỏi 1: Hạng vé nào có tỷ lệ sống sót cao nhất, chênh lệch bao nhiêu so với hạng thấp nhất?

In [15]:
# TODO
pclass_survival = df.groupby('pclass')['survived'].mean()
print("Tỷ lệ sống sót theo hạng vé:")
print(pclass_survival)

diff = pclass_survival.max() - pclass_survival.min()
print(f"\nHạng {pclass_survival.idxmax()} cao nhất ({pclass_survival.max():.2%}) ")
print(f"=> Chênh lệch: {diff * 100:.2f}%")

Tỷ lệ sống sót theo hạng vé:
pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64

Hạng 1 cao nhất (62.96%) 
=> Chênh lệch: 38.73%


## Câu hỏi 2: Giới tính hay hạng vé ảnh hưởng đến sống sót mạnh hơn?

In [17]:
# TODO
sex_survival = df.groupby('sex')['survived'].mean()
sex_diff = sex_survival.max() - sex_survival.min()

pclass_survival = df.groupby('pclass')['survived'].mean()
pclass_diff = pclass_survival.max() - pclass_survival.min()

print("Tỷ lệ sống sót theo giới tính:")
print(sex_survival)
print(f"-> Chênh lệch theo giới tính: {sex_diff * 100:.2f}%\n")

print("Tỷ lệ sống sót theo hạng vé:")
print(pclass_survival)
print(f"-> Chênh lệch theo hạng vé: {pclass_diff * 100:.2f}%\n")


Tỷ lệ sống sót theo giới tính:
sex
female    0.742038
male      0.188908
Name: survived, dtype: float64
-> Chênh lệch theo giới tính: 55.31%

Tỷ lệ sống sót theo hạng vé:
pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64
-> Chênh lệch theo hạng vé: 38.73%



## Câu hỏi 3: Vé đắt hơn có thực sự sống sót cao hơn không?

In [20]:
# TODO
print("Giá vé trung bình theo tình trạng sống sót:")
print(df.groupby('survived')['fare'].mean())

print(f"\nHệ số tương quan (fare & survived): {df['fare'].corr(df['survived']):.4f}")


Giá vé trung bình theo tình trạng sống sót:
survived
0    22.117887
1    48.395408
Name: fare, dtype: float64

Hệ số tương quan (fare & survived): 0.2573


## Câu hỏi 4: Gia đình đông người có ảnh hưởng đến khả năng sống sót không?

In [22]:
# TODO
family_survival = df.groupby('family_size')['survived'].agg(['count', 'mean'])
display(family_survival)

,count,mean
family_size,,
1,537,0.303538
2,161,0.552795
3,102,0.578431
4,29,0.724138
5,15,0.200000
6,22,0.136364
7,12,0.333333
8,6,0.000000
11,7,0.000000


## Câu hỏi 5: Cảng lên tàu (embark_town) nào có tỷ lệ sống sót cao nhất?

In [23]:
# TODO
embark_survival = df.groupby('embark_town')['survived'].agg(['count', 'mean']).sort_values(by='mean', ascending=False)
display(embark_survival)

,count,mean
embark_town,,
Cherbourg,168,0.553571
Queenstown,77,0.389610
Southampton,644,0.336957


## Câu hỏi 6 (Tổng hợp) — Viết insight tổng hợp
Dựa trên Phần A, B, C, viết 4-5 câu insight tổng thể về dữ liệu Titanic.

Giới tính là yếu tố quyết định mạnh nhất: Nữ giới có tỷ lệ sống sót vượt trội so với nam giới (khoảng 74% so với ~19%), phản ánh chính sách ưu tiên phụ nữ khi sơ tán.  
Hạng vé (pclass) và giá vé (fare) tỷ lệ thuận với cơ hội sống sót: Hành khách ở Hạng 1 (First Class) và nhóm mua vé giá cao có tỷ lệ sống sót cao nhất, giảm dần về phía Hạng 3.  
Quy mô gia đình lý tưởng: Đi theo nhóm nhỏ (2–4 người) có tỷ lệ sống sót cao nhất; trong khi đi một mình hoặc đi cùng gia đình quá đông (>4 người) khiến tỷ lệ sống sót giảm mạnh.  
Cảng Cherbourg ghi nhận tỷ lệ sống sót cao nhất: Nguyên nhân chính là do tỷ lệ hành khách mua vé Hạng 1 lên tàu tại cảng này chiếm tỷ trọng lớn hơn các cảng khác.  